### ISSUE TITLE
How to display a list with duplicate strings

### ISSUE THREAD

**Issue by bender2k14** [elmish/Elmish.WPF#131]
I want to populate a [`ListBox`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.listbox?view=netframework-4.8) with a list of strings that are _not necessarily distinct_ (as strings).

Here is the binding I used (after I added C#-style name arguments for clarity here even though it doesn't compile)...
```
Binding.oneWaySeq(
  get: (fun m -> m.MyListOfStrings),
  itemEquals: refEq,
  getId: id)
```

...which calls this method:
https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/Binding.fs#L362-L365

Even though I will have structurally equal strings, I expected this to work because
1. the list of string instances are stored in my model,
2. I used the reference equality function `refeq` to check for the equality of two string instances, and
3. the strings are all constructed at runtime, so they are not reference equal via [string interning](https://docs.microsoft.com/en-us/dotnet/api/system.string.intern?view=netframework-4.8).

Now that my use case has brought my attention here, I find this a bit problematic.  The issue is how instances of type `'a` are checked for equality.  The obvious choice is `itemEquals`, but it is also possible to check for equality after mapping to instances of type `'id`.  The documentation says that `getId` should return a unique ID for each item...
https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/Binding.fs#L349-L361

...but it doesn't say how those instances of type `'id` are checked for equality.

This come to a head in this code.
https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/ViewModel.fs#L332-L339

In my case, the code throws an `ArgumentException` from `newLookup.Add(getId v, v)` on line 337 when trying to add a duplicate string (as specified in the documentation for [`Dictionary<K,V>.Add(K, V)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.add?view=netframework-4.8)) since the [default `Dictionary<,>` constructor](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.-ctor?view=netframework-4.8#System_Collections_Generic_Dictionary_2__ctor) checks for equality of its keys using the [default equality comparer](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.equalitycomparer-1.default?view=netframework-4.8), which is a structural equality for the type `string`.

As I see it, the reason for using a dictionary here is to get the `newVal` corresponding to `existingVal` (or determine that no such value exists) in effectively constant time so that the combined runtime with the `for` loop on the previous line, 336, is effectively linear.  The proper way to achieve effective constant time is to use a hash function instead of an "ID function".  This is the kind of thing I was trying to explain in #117.  However, achieving linear-time execution for that part of the code isn't much of an improvement given that the code just below...
https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/ViewModel.fs#L346-L350

...executes in quadratic time.

The naive alternative is to replace the `match` on line 339 (and the cases after it) with something like...
```
match newVals |> Seq.filter (itemEquals existingVal) |> Seq.length with
| 0 -> // remove existingVal logic
| 1 -> () // keep existing value
| _ -> failwith "itemEquals found duplicate equal new values"
```

...which would yield an overall quadratic runtime (for this part of the code).  If I make that change and also replace the [other](https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/ViewModel.fs#L349) [two](https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/ViewModel.fs#L357) equality checks for instances of type '`a` via structural equality checks on mapped instances of type `'id` (via F#'s `=` operator) so that `getId` is not longer used, then my code behaves as I expect.

I created the branch [bender2k14/Elmish.WPF/tree/string_list_with_duplicates](https://github.com/bender2k14/Elmish.WPF/tree/string_list_with_duplicates) with two commits.  The first one modifies the `OneWaySeq` sample to use strings and reproduces the problem.  The second commit makes the changes that I described above to obtain the expected behavior.

In conclusion, I recommend the following:
1. only check for equality of two instances of a user-defined type with a user-supplied equality function;
2. if it is desired to reduce the runtime of these nested loops from quadratic time to effectively linear time, then use a user supplied hash function.

In C#/.NET, the pair of hash code and equality functions is commonly expressed via the interface [`IEqualityComparer<>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1?view=netframework-4.8).

I think I understand this `OneWaySeq` case well enough to say that we should take my first recommendation by replacing uses of `getId` with `itemEquals` like I did in my second commit.  As for the second recommendation, if we want to take that as well, then we would take in a hash function instead of `getId`, use it and `itemEquals` to create an instance of `IEqualityComparer<>`, and then create that dictionary using [this constructor overload](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.-ctor?view=netframework-4.8#System_Collections_Generic_Dictionary_2__ctor_System_Collections_Generic_IEqualityComparer__0__).

However, I don't think we should just merge in my second commit right away though.  I want to study the `OneWaySeq` case in the `updateValue` function some more, and we should also test reordering and removing items (instead of just adding them).

**Comment by BentTranberg** [comment]
I may be in over my head here as far as understanding the problem, and I don't want to participate in judging which solution is best, again due to lack of understanding. I just want to point out one particular possibility. The user can make the decision whether to deduplicate string data, either through use of Intern, or through own code. E.g. in C# one might simply `if (s1 = s2) s1 = s2;` to deduplicate. Using Intern can be a bit unpredictable as far as performance is concerned, but manually deduplicating e.g. a list of strings read from a database is considerably more predictable. I have done this many times in order to save memory and/or improve performance.

**Comment by bender2k14** [comment]
Those are ways to remove duplicates from a list. Another is calling [`Seq.distinct`](https://msdn.microsoft.com/visualfsharpdocs/conceptual/seq.distinct%5b%27t%5d-function-%5bfsharp%5d).

However, I am saying that I do want structurally equal duplicates (but not reference equal duplicates).  I don't want them removed.

**Comment by BentTranberg** [comment]
We are talking about the same thing, @bender2k14.

The C# statement I used as an example - `if (s1 = s2) s1 = s2;` is completely meaningless, unless you take into account that it has the effect of making the two string share the string data on the heap - or as you say, have reference equality. This particular example might only have a benefit if the two strings were really long, or a string comparison was done very frequently on the two.

But the same manual technique can be used if you read a string field from a database, and you know beforehand that this field will likely have lots and lots of rather specific duplicates - e.g. "John Smith" or "Initial State". Then you would read the field from DB, go through your (possibly temporary) list/collection/whatever for that field in search of an equal string, and then make the two referencial equal too. Memory saved, future performance increased. This is faster and more predictable than using Intern, which has to consider all your strings.

My point is that if any suggested solution has a downside, then it might be better to just let the user deal with it, e.g. in this way I've described, rather than build something into the library.

But your suggested solution, as far as I can see, appears to have no downside.

**Comment by cmeeren** [comment]
This issue went a bit above my head on the first reading. I'll have to get back to it when I have the time to delve into it.

I do have two questions, though:
* `oneWaySeq` is (perhaps naïvely) only intended for distinct items (as returned by the `getId` function). Are you sure it's the right binding for the job? Might another existing one, or a new one, fit better?
* Have you tried using `LanguagePrimitives.PhysicalHash` as `getId`? Does that solve your problem?

**Comment by bender2k14** [comment]
> This issue went a bit above my head on the first reading.
> ...
> I do have two questions, though...

I could be making a mistake in my example, and either of your suggestions might be the right away to go about it.  After I have thoroughly investigated your questions @cmeeren and your response @BentTranberg, I will reply.

> It should be documented that it uses the default equality comparison `(=)`, like a few other bindings do. That was the intention.

More than that actually.  For...
```
getId: 'a -> 'id
itemEquals: 'a -> 'a -> bool
```
... I think the requirement is
> For all `x`, `y`, of type `'a`, it follows that `(itemEquals x y) = ((getId x) = (getId y))`

---Aside---
> I could be making a mistake in my example...

In fact, I was.  Passing in `id` for `getId` and `refEq` for `itemEquals` with `'a` equal to `string` does not satisfy this requirement.  Your suggestion to pass in `LanguagePrimitives.PhysicalHash` for `getId` is much more likely to satisfy that requirement, but it appears that [garbage collector can defragment (part of) the heap](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.gclargeobjectheapcompactionmode?view=netframework-4.8#System_Runtime_GCLargeObjectHeapCompactionMode_CompactOnce), so this isn't a valid hash function.  (A valid hash function needs to return the same value for the same instance for the lifetime of the instance.)  Anyway, my two-second attempt using `LanguagePrimitives.PhysicalHash` didn't work either, but I need to look at it more carefully when I have the time.
---End Aside---

It can be difficult enough to properly define one equality function.  In addition to being reflexive, symmetric, and transitive, this equality functions `itemEquals` and `fun x y -> (getId x) = (getId y)` must evaluate to `false` for every distinct pair of items in `vals` (respectively, `newVals`), where distinct here means at distinct positions in the `ObservableCollection<>` (respectively, new `seq<>`).  For a given equality function, let's call this additional condition the uniqueness condition.

In place of requiring the user to supply both `getId` and `itemEquals`, here are common alternatives that I have seen used in similar situations.
1. Just `itemEquals: 'a -> 'a -> bool`, which is an equality function and satisfies the uniqueness condition.  The runtime would be quadratic.
2. Just `getId: 'a -> 'id`, which defines an equality function `fun x y -> (getId x) = (getId y)` that satisfies the uniqueness condition.  The runtime could be nearly linear time (if we implement `updateValue` efficiently).
3. `itemEquals: 'a -> 'a -> bool` and `itemHash: 'a -> int`, where `itemEquals` is an equality function, `itemEquals` satisfies the uniqueness condition, and for all `x` and `y` of type `'a`, if `itemEquals x y`, then `(itemHash x) = (itemHash y)`.  The runtime could be nearly linear time (if we implement `updateValue` efficiently and `itemHash` is a "good" hash function).

The requirements of each alternative from weakest to strongest are 1, 3, 2.  Clearly 1 is weaker than 3.  Then 3 is weaker than 2 because `getId` can be passed in for `itemHash` (never a collision...a great hash function) and `fun x y -> (getId x) = (getId y)` can be passed it for `itemEquals`.

I am not overly concerned with efficiency.  The input size has to be rather large before the difference between nearly linear and quadratic runtimes will be noticeable to the user, and this "input" is supposed to be put into an `ObservableCollection<>` in the GUI.  If someone is trying to display that much data in the GUI, then I think they have more fundamental issues to worry about than quadratic update times.  This is one of the reasons why the second commit in my branch (https://github.com/bender2k14/Elmish.WPF/commit/8fb9808bbfb5d682e44ee7540b71d8971363b947) implements 1, the simplest of these three alternatives.

**Comment by bender2k14** [comment]
> It should be documented that it uses the default equality comparison (=), like a few other bindings do. That was the intention.

And it would be great if the corresponding type constraint, `requires equality`, were on `'id`, especially if the type system was able to infer this requirement.  I think the use of boxing currently prevents the type system from realizing this though.

**Comment by BentTranberg** [comment]
Thanks for all the work you two are doing lately. It's great. I've had no problem with the updates, and the instructions for update was great. The following is, I guess, slightly off subject, but hopefully of interest.

I have experience with Delphi since it's inception way back in the eighties, and with .NET (C# and later F#) since 2008. The mechanism for strings is the same in these languages/platforms. I've built backend and frontend applications where massive memory and CPU use was an issue to be dealt with. Huge and complex memory structures doesn't necessarily eat up the machines memory, but it can seriously affect the heap and CPU. (Comparing massive numbers of equal strings that don't share data will eat CPU.) What I've noticed is that strings tend to be the first I look at in my profiler tool. Other data and types can be dealt with in specific other ways, sometimes through design patterns. Strings however, is in a class of its own. It tends to be the most used type, and the most used object - at least for my kind of apps. It is a very general data type, It has a very specific and unique memory handling mechanism that separates it completely from other basic types.

That's why I brought up the subject of string deduplication - one of my favorite tools for preventing and fixing memory problems, and sometimes increase performance. I've used this technique to be able to squeeze huge amounts of data into memory that would otherwise definitely not fit, or would otherwise cause deadly fragmentation on the heap. I've used it to get rid of the worst noise in the profiler statistics, so that other memory issues could surface. I've used it with GUI components - e.g. virtual list boxes - to speed up things and avoid out of memory, back in the Delphi days. I could maybe mention more, but I have to run to my niece's wedding!

**Comment by cmeeren** [comment]
A couple further notes:

`oneWaySeq` is, as I said, intended for distinct items. Just want to add that this is specified in the docs (emphasis mine):

> Creates a one-way binding to a sequence of items, each **uniquely** identified by the value returned by `getId`.

It could definitely be made more clear, e.g. "to a distinct sequence".

---

> And it would be great if the corresponding type constraint, `requires equality`, were on `'id`

I agree.

---

@bender2k14 In general, this is perhaps just a bit too technically oriented for me. I understand most of what you say, but I am coming at this (and library design in general) from a whole other angle, centring on these questions:

1. Which problem is the user struggling with?
2. How can the API/functionality be improved to solve that problem?
3. API ergonomics - how can we implement the desired functionality while placing minimal requirements on the user (for example, what is the minimal, useful set of parameters to require the user to supply)?

In light of the above questions, `oneWaySeq` binding can be answered as follows:

1. The user wants to display a long list of "simple" values that are bound to directly from the UI (for complex values with separate Elmish.WPF-defined bindings for each item, there is `subModelSeq`). Note that this is not limited to primitive types. For any given update loop, at most one or a few values are changed/reordered/etc. Let's say rendering the whole list locks the UI for a duration on the order of seconds. If using `oneWay`, the entire list is re-rendered for every model update, which is prohibitively expensive. If using `oneWayLazy`, the list is only re-rendered when it's updated, which may still be prohibitively expensive, and it's also unnecessary because it "should" be possible to only re-render what has changed.
2. Create a binding for a list of simple values, where only changed items are re-rendered.
3. (As I understand it, this is the core of this particular issue.) In order to ensure only changed values are re-rendered, we must be able to
    * know which values in the old list correspond to which values in the new list (items may have been re-ordered, added, or removed, so there might not be a correspondence), and
    * once we have found two corresponding values, determine whether they are equal.

   In order to determine which items to compare, we must assume that the items can be distinctly identified. We therefore need a function that accepts an item and returns a unique identifier. This is not necessarily a hash function, because the items might be e.g. a record type (you could bind to its properties from the XAML) with an `Id` property for identity and other values that might change.

    In order to determine if two items are equal, we can use a function that accepts two items and returns a boolean indicating equality. (Another possibility is a user-supplied hash function, though at the time I either didn't think about it, or I didn't know how that might impact the effectiveness of the implementation and/or figured that a comparison function was simpler to understand.)

From the above considerations, the current `oneWaySeq` API falls out quite naturally, along with the aforementioned uniqueness constraint in the function's docs.

https://github.com/elmish/Elmish.WPF/blob/10b4deb160cbe1808e4d2f68ebd9c772e48d0ecc/src/Elmish.WPF/Binding.fs#L362-L365

Let me finally add that the initial problem, or at least the specific variant described above, might of course be something very few users are struggling with, rendering the whole point moot. And the assumptions I have made may of course be false.

**Comment by bender2k14** [comment]
Thanks very much @cmeeren sharing that detailed comment!  It has helped my clarify my mind in two key things that were confusing me.

> ...the items might be e.g. a record type (you could bind to its properties from the XAML) with an Id property for identity and other values that might change.

First, I think I better understand the initial problem I had trying to display a list of strings.  I forgot that in GUI programming there is often more data in the view than is displayed to the user.  If I only want to display a list of strings to the user, then I can put into the view a record like `{ int: Id; data: string }`.  Something like this is often necessary because when the user wants to change/remove one of the strings, we have to know which one should be changed/removed.  When I wanted to display a list of strings to the user, it was sort of as a minimal working example, so I didn't try to implement any way to change/remove these strings or add additional data (like an ID) to support such an implementation.  I also wasn't completely sure how to format this string.  You said here that this can be done in the XAML.  That sounds great to me.  This would be done using [ItemTemplate](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.itemscontrol.itemtemplate?view=netframework-4.8), right?  I think so.  I try this tomorrow at work.  And to be clear, this first/initial confusion of mine doesn't have much to do with anything that follows.  It was just that I noticed these other things while trying to solve this first/initial problem.

-----------

Second, I figured out why I was so confused about the roles of `getId` and `itemEquals`.  I suggested that they would essentially play the same role in the sense that
> For all `x`, `y`, of type `'a`, it follows that `(itemEquals x y) = ((getId x) = (getId y))`

I said that even knowing that the code is consistent with your description that `getId` is used to find one item from each list that "match", and `itemEquals` is used to check if those two are "equal".  Specifically, given two items `x` and `y` that yield the same `'id` from `getId`, sometimes `itemEquals x y` evaluates to `true` and sometimes it evaluates to `false` (and this is by design).

In short, I realized that `itemEquals` is part of the "update" logic; it is not used to find pairs of items (one from each list) that "match".

Now to expand on that with a longer answer.

> 3. (As I understand it, this is the core of this particular issue.)

Indeed; this is the core of the "second" issue that caught my attending while working on the "first" one.

> In order to ensure only changed values are re-rendered, we must be able to
>    - know which values in the old list correspond to which values in the new list (items may have been re-ordered, added, or removed, so there might not be a correspondence), and

Yep.  Let's call such a corresponding pair a match.

>    - once we have found two corresponding values, determine whether they are equal.

This was one of the things that I found so confusing.  Instead of "determine whether _they are equal_", it would be more precise to say "determine whether _to keep the old item_" (instead of replacing it with the new item).  It might typically be the case that the old and new items not being (say, structurally) equal means that the old item should be replaced with the new item, but that is not necessary.

For example, if the item type is the record...
```
type MyRecord = { Id: int; Value: string; Foo: int }
```
...and only the `Value` field is displayed to the user, then just because...
```
let oldItem = { Id = 0; Value = "Bar"; Bar = 0 }
let newItem = { Id = 0; Value = "Bar"; Bar = 1 }
```
...are not (structurally) equal as instances of type `MyRecord` doesn't mean `newItem` should replace `oldItem`.  It is up to the user to decide whether this replacement should occur.

So instead of calling the function `itemEquals`, it makes more sense to me to call it something like `keepOld`.

> In order to determine which items to compare, we must assume that the items can be distinctly identified. We therefore need a function that accepts an item and returns a unique identifier.

No, it is not necessary, but it is sufficient.  In order to determine which items match, it suffices to have a function `itemMatch: 'a -> 'a -> bool`.  The matches are
```
let matches =
  seq {
    for newVal in newVals do
      for oldVal in vals do
        if itemMatch newVal oldVal
        then yield (newVal, oldVal)
  }
```

If the items can be distinctly identified (say) by a function `getId: 'a -> 'id`, then we can use `getId` to define `itemMatch` via
```
let itemMatch getId x y =
  (getId x) = (getId y)
```

----------

My company has a library containing code we reuse on most of our client projects.  One of those functions that we often reuse is called `Merge`, which "merges" a new sequence of values into an old sequence of values.  It was named after the corresponding SQL statement.  Here is the [Microsoft documentation](https://docs.microsoft.com/en-us/sql/t-sql/statements/merge-transact-sql?view=sql-server-2017) for that statement.  I like the syntax given for this statement that I found [here](https://www.mssqltips.com/sqlservertip/1704/using-merge-in-sql-server-to-insert-update-and-delete-at-the-same-time/), which (after some small improvements) looks like
```
MERGE <target_table> [AS TARGET]
USING <table_source> [AS SOURCE]
ON <search_condition>
[WHEN MATCHED 
   THEN <merge_matched> ]
[WHEN NOT MATCHED [BY TARGET]
   THEN <merge_not_matched_by_target> ]
[WHEN NOT MATCHED [BY SOURCE]
   THEN <merge_no_matched_by_source> ];
```

The API of my company's `merge` function is similar.  It looks like
```
let merge
  (source: seq<'s>)
  (target: seq<'t>)
  (match: 's -> 't -> bool)
  (create: 's -> 't)
  (add: 's -> unit)
  (update: 's -> 't -> unit)
  (remove: 's -> unit)
  : unit =
// implementation
```

If we were to implement this `OneWaySeq` case of `updateValue` using this `merge` function (in the most direct way), it would look something like
```
let source = newVals
let target = vals
let match x y = (getId x) = (getId y)
let create = id
let add = vals.Add
let update newItem oldItem =
  if not <| itemKeepOld newItem OldItem then 
    vals.Remove oldVal |> ignore
    vals.Add newVal
let remove = vals.Remove >> ignore

merge source target match create add update remove
```

(Although, my company's 'merge' function doesn't concern itself with ordering, so that would have to be handled separately.)

Aside:
My company has many variants of this `merge` function.  For some of them, an argument is `match: 's -> 't -> bool` as above while others replace this `match` function with ID functions `getSourceId: 's -> 'id` and `getTargetId: 't -> 'id`.  Not only is this easier for the consumer to call if such IDs exist, but it also permits a more efficient implement of `merge` (with a running time near linear instead of quadratic).

--------

Now, let's look closer at this update logic.  The update logic in the `OneWaySeq` case is
https://github.com/elmish/Elmish.WPF/blob/8f78f833bef5c903e6947b23688cef30480ccf8b/src/Elmish.WPF/ViewModel.fs#L341-L343

For comparison, the update logic in the very similar `SubModelSeq` case is
https://github.com/elmish/Elmish.WPF/blob/8f78f833bef5c903e6947b23688cef30480ccf8b/src/Elmish.WPF/ViewModel.fs#L461

The `SubModelSeq` case is completely general in that the update case calls a function called `update` that does whatever it needs to do to update.  Looking back at the `OneWaySeq` case, the logic in the update case is very specific.  In fact, it is exactly the `add` and `remove` logic after getting though some `if` check.  But that `if` check is also very specific.  There is a sense in which it has the role as `match`.

What I am getting at is that another way to implement this `OneWaySeq` case of `updateValue` using this `merge` function would look something like
```
let source = newVals
let target = vals
let match newVal oldVal = ((getId newVal) = (getId oldVal)) && (keepOld newVal oldVal)
let create = id
let add = vals.Add
let update _ _ = ()
let remove = vals.Remove >> ignore

merge source target match create add update remove
```

This is what I meant in previous comments when I said that we don't need both `getId` and `itemEquals`.  It suffices to just have one function of type `'a -> 'a -> bool`.  I now see why saying "we don't need `getId`, just `itemEquals`", while logically correct, was (and maybe still is) conceptually confusing.

-------------

P.S.
While writing this comment, I think I found a separate but related issue.  Recall the update logic for the `OneWaySeq` case of `updateValue`.
https://github.com/elmish/Elmish.WPF/blob/8f78f833bef5c903e6947b23688cef30480ccf8b/src/Elmish.WPF/ViewModel.fs#L341-L343

I think that call to `Remove` contains a bug.  The precise thing to do would be to remove by the index.  Removing by the item means that the collection needs to find which item to remove.  The documentation for `ObservableCollection<T>.Remove` [says](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1.remove?view=netframework-4.8#remarks) it removes the item it finds for which that item and the given item return `true` when passed into the default equality comparer of type `T`.  If `itemEquals` is not the same as this default equality comparer, then this call to `Remove` will not have the intended behavior.  I can create a test that exhibits this problem and a PR with a fix.

**Comment by bender2k14** [comment]
> In short, I realized that `itemEquals` is part of the "update" logic; it is not used to find pairs of items (one from each list) that "match".

Therefore, I am not longer suggesting that we _should_ change the API of `OneWaySeq`.  Although it is not necessary for the items to have an ID, I do think it is reasonable that the API for `OneWaySeq` requires the items to have an ID.  An ID is
- convenient for us,
- convenient for users,
- easy for users to make item types with an ID, and
- I expect that users would already have an ID on their item type for other reason.*

*Even though I didn't have an ID on my item type.  I was just trying to some something simple, but I accidently did something so simple that it was unusual.

> `oneWaySeq` is (perhaps naïvely) only intended for distinct items (as returned by the `getId` function). Are you sure it's the right binding for the job? Might another existing one, or a new one, fit better?

Yes, it is reasonable to consider adding a slightly different API for creating a `OneWaySeq`.  Recall that my company has different versions of `merge`: some have `match: 's -> 't -> bool` as an argument while others replace this `match` function with ID functions `getSourceId: 's -> 'id` and `getTargetId: 't -> 'id`.  In the same way, we could add another way to create a `OneWaySeq` binding given just `get: 'model -> #seq<'a>` and `match: 'a -> 'a -> bool`.

I don't care if we add this addition binding API now, when someone requests it, or never.  However, I do think it is important to realize that we have this choice.  The runtime of this case in `updateValue` would be "quadratic", or essentially the same as the current `OneWaySeq` case is now.  Which is a shame, because with IDs for every item, the runtime can be nearly linear.  I will create PRs that make full use of the IDs and achieve nearly linear runtime for the current `OneWaySeq` case in `updateValue`.

**Comment by cmeeren** [comment]
Thank you for the detailed explanation! I agree with (I think) everything you have said in your two last comments, and appreciate the insights you have shared. I am looking forward to your improvements in the `update` implementation. You may notice that there are several branches where an `ObservableCollection` is updated; if there is a way to reconcile these (without losing performance or adding unnecessary complexity), that's very welcome too (but certainly not required).

I also have no problems with additional overloads for `oneWaySeq` (and perhaps also `oneWaySeqLazy`?), though, as you, I don't care if this is added now, upon request, or never. I personally don't need it at the moment.

Feel free to close this issue if you think it's done, and thanks again for being so thorough!

**Comment by bender2k14** [comment]
> Thanks for all the work you two are doing lately. It's great.

Thanks @BentTranberg!  I am having lots of fun working on this :)

> I have experience with Delphi since it's inception way back in the eighties...

My hat is off to you.  That is so much programming experience.  I have been a professional programmer for four years now and programmed as a student for the 12 years before that.  It blows my mind to think that it is reasonably likely that I will still be in this profession when I am twice my current age!

> What I've noticed is that strings tend to be the first I look at in my profiler tool.

I have investigated and fixed memory leaks before, but I have yet to make a profiler a standard tool of mine.  It not for lack of desire.  I think the main problem is me not understanding the profiler's data.  If I did, then I would expect that I could more easily drill into the data, become informed about specific behavior of the code, and make a change to the code with confidence about how that change will impact the memory usage.

> ...read the field from DB, go through your (possibly temporary) list/collection/whatever for that field in search of an equal string, and then make the two referencial equal too. Memory saved, future performance increased.

Now that you mention it, I recall reading a story about how some application was optimized.  The main improvement was turning some object into a singleton and reusing that same instance everywhere instead of repeatedly creating new temporary copies.

> I've used it to get rid of the worst noise in the profiler statistics, so that other memory issues could surface.

Oh, that is a really good idea.  I hope I can remember that when I eventually become proficient with a profiler.

**Comment by bender2k14** [comment]
> ...thanks again for being so thorough!

You are welcome!  I enjoy it :)

> Feel free to close this issue if you think it's done...

Sure.  In closing, here is a summary of this issue.

I opened this issue when I had difficulty displaying a list with duplicate strings.  Instead of providing a list of `string`s to a `OneWaySeq` binding, I needed to provide a list containing a type like
```
type MyRecord { Id: Guid; Data: string }
```

Then in the XAML, I can format the display the of each item in a `DataTemplate` like
```
<ListBox ItemsSource="{Binding MyStrings}">
  <ListBox.ItemTemplate>
    <DataTemplate>
      <TextBlock Text="{Binding Data, StringFormat='My string is {0}'}" />
    </DataTemplate>
  </ListBox.ItemTemplate>
</ListBox>
```

Now that I look closer, the `SubModelSelectedItem` sample uses a `DataTemplate` like this.
https://github.com/elmish/Elmish.WPF/blob/852298b83f1a1920376271385d11d06f332c3711/src/Samples/SubModelSelectedItem.Views/MainWindow.xaml#L13-L20

FWIW, I found another way to write that part in the middle.  It can also be expressed as
```
<TextBlock>
  <Run Text="{Binding Name, Mode=OneWay}"/>
  <Run Text="{Binding SelectedLabel, Mode=OneWay}"/>
</TextBlock>
```

Anyway, during the discussion, three other issues were raised.  They now have GitHub issues of their own (#134, #137, #139).

**Comment by bender2k14** [comment]
> ...for complex values with separate Elmish.WPF-defined bindings for each item, there is `subModelSeq`

I just realized that `subModelSeq` can be used to specify "sub-model" bindings without also specifying other things like its own message type.  The `SubModelSelectedItem` sample does that here
https://github.com/elmish/Elmish.WPF/blob/66a6b49978c10f3a9c40c479ca8957517acba0ff/src/Samples/SubModelSelectedItem/Program.fs#L32-L38

In contrast, I was using a `OneWaySeq` binding like this
```
// in F#
type IdString { Id: Guid; Data: string }
type Model { IdStrings: IdString list }

let bindings () = [
  "IdStringsBindingName" |> Binding.oneWaySeq(
    (fun m -> m.Messages),
    (fun m -> m.Data),
    (fun m -> m.Id))

// in XAML
<ListBox ItemsSource="{Binding IdStringsBindingName}">
  <ListBox.ItemTemplate>
    <DataTemplate>
      <TextBlock Text="{Binding Data}" />
    </DataTemplate>
  </ListBox.ItemTemplate>
</ListBox>
```

This means that the field name "Data" in "IdString.Data" is now included in the set of magic GUI binding strings.  Before I added code like that to my project, all the magic GUI binding strings were like `"IdStringsBindingName"` in that there were all string literals passed into an Elmish.WPF binding, which in this case looks like `"IdStringsBindingName" |> Binding.oneWaySeq(...)`.  This means that I can now "break" my application so that it only (mildly) fails at runtime by doing Refactor-Rename in Visual Studio on the name of that "Data" field.  **YIKES!**

To compensate for this, I already made the following change.  I added another type called `IdStringViewModel` with the same structure as `IdString`, defined it just before the bindings, and mapped the elements in `IdStrings` to this new type.  That way, at least the set of magic strings were "closer" together in the F# code (and in my solution, I have the model types in a different project than the bindings, so this changes is more significant in that case than it seems when applied to the above example).  This is what I used to do in previous work projects when the C# property or field names are also magic strings that need to match something elsewhere in the application.  That way, I know I can use Visual Studio's Refactor-Rename feature on any code except for that code defined at such boundaries.

I _MUCH_ prefer how the `SubModelSelectedItem` sample does this.  I will be changing my code at work to be like that soon.
